crates/symbolicator/src/services/symbolication.rs

```diff
@@ -130,76 +130,99 @@ impl<'a> SymCacheLookupResult<'a> {
+    fn has_module(&self, id: &CodeModuleId) -> bool {
+        self.inner.contains_key(id)
+    }
+
+    /// Extend the CacheModules with the fetched caches represented by
+    /// [`CfiCacheResult`].
+    fn extend(&mut self, cfi_caches: Vec<CfiCacheResult>) {
+        // make sure to hold onto a reference to the CfiCacheFile, to make sure
+        // it will not be evicted in the middle of using it
+        self.cache_files.extend(
+            cfi_caches
+                .iter()
+                .filter_map(|(_, cache_result)| cache_result.as_ref().ok())
+                .map(Arc::clone),
+        );
+
+        let iter = cfi_caches.into_iter().map(|(code_id, cache_result)| {
+            let cfi_module = match cache_result {
+                Ok(cfi_cache) => {
+                    let cfi_status = match cfi_cache.status() {
+                        CacheStatus::Positive => ObjectFileStatus::Found,
+                        CacheStatus::Negative => ObjectFileStatus::Missing,
+                        CacheStatus::Malformed => {
+                            let err = CfiCacheError::ObjectParsing(ObjectError::Malformed);
+                            log::warn!("Error while parsing cficache: {}", LogError(&err));
+                            ObjectFileStatus::from(&err)
                         }
+                    };
+                    let cfi_path = match cfi_cache.status() {
+                        CacheStatus::Positive => Some(cfi_cache.path().to_owned()),
+                        _ => None,
+                    };
+                    CfiModule {
+                        features: cfi_cache.features(),
+                        cfi_status,
+                        cfi_path,
+                        cfi_candidates: cfi_cache.candidates().clone(), // TODO(flub): fix clone
+                        ..Default::default()
                     }
-                    Err(err) => {
-                        log::debug!("Error while fetching cficache: {}", LogError(err.as_ref()));
-                        CfiModule {
-                            cfi_status: ObjectFileStatus::from(err.as_ref()),
-                            ..Default::default()
-                        }
+                }
+                Err(err) => {
+                    log::debug!("Error while fetching cficache: {}", LogError(err.as_ref()));
+                    CfiModule {
+                        cfi_status: ObjectFileStatus::from(err.as_ref()),
+                        ..Default::default()
                     }
-                };
-                (*code_id, cfi_module)
-            })
-            .collect();
-        Self { inner }
+                }
+            };
+            (code_id, cfi_module)
+        });
+        self.inner.extend(iter)
+    }
+
+    /// Returns a copy of the inner Module Map that can be used for processing.
+    fn for_processing(&self) -> BTreeMap<CodeModuleId, CfiModule> {
+        self.inner.clone()
```

Comment:
cloning a `BTreeMap` sounds expensive?

Reply:
This is indeed unfortunate. Previously, the `CfiCacheModules` (which internally was a vec of tuples) would be serialized and sent via procspawn. Now it is the btreemap, so not too different actually.